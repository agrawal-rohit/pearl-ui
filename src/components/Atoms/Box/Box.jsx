import React from "react";
import { View } from "react-native";
import useStyledProps from "../../../hooks/useStyledProps";
import { useTheme } from "../../../hooks/useTheme";
import {
  backgroundColor,
  border,
  layout,
  opacity,
  position,
  shadow,
  spacing,
  visible,
} from "../../../theme/src/styleFunctions";
import {
  IBasePearlTheme,
  RestyleFunctionContainer,
} from "../../../theme/src/types";

// export type BoxProps<
//   Theme extends IBasePearlTheme
// > = BackgroundColorProps<Theme> &
//   OpacityProps &
//   VisibleProps &
//   LayoutProps &
//   SpacingProps<Theme> &
//   BorderProps<Theme> &
//   ShadowProps<Theme> &
//   PositionProps<Theme>;

export const boxRestyleFunctions = [
  backgroundColor,
  opacity,
  visible,
  layout,
  spacing,
  border,
  shadow,
  position,
];

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built. By default, it renders a <View> element.
 */
// type ViewProps = React.ComponentProps<typeof View> & {
//   children?: React.ReactNode;
// };
// type ComponentProps<Theme extends IBasePearlTheme> = BoxProps<Theme> &
//   Omit<ViewProps, keyof BoxProps<Theme>>;

const Box = React.forwardRef((props, ref) => {
  const passedProps = useStyledProps(boxRestyleFunctions, props);

  return <View ref={ref} {...passedProps} />;
});

export default Box;
