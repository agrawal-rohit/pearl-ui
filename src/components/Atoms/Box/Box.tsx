import React from "react";
import { View, ViewProps } from "react-native";
import { pearlify } from "../../../pearlify";

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built.
 * By default, it renders a <View> element.
 */
const Box = pearlify<ViewProps>(View, {
  componentName: "Box",
  type: "basic",
  animatable: true,
});

// The props that the Box component accepts.
export type BoxProps = React.ComponentProps<typeof Box>;

// Export the Box component as the default export.
export default Box;
