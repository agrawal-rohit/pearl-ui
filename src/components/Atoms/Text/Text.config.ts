export default {
  baseStyle: {
    color: {
      light: "neutral.900",
      dark: "neutral.50",
    },
    scaleFontSize: true,
  },
  variants: {
    h1: {
      fontFamily: "heading",
      fontWeight: "bold",
      fontSize: "8xl",
      lineHeight: "10xl",
    },
    h2: {
      fontFamily: "heading",
      fontWeight: "bold",
      fontSize: "5xl",
      lineHeight: "7xl",
    },
    t1: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "2xl",
      lineHeight: "4xl",
    },
    t2: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "l",
      lineHeight: "2xl",
    },
    st1: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "m",
      lineHeight: "xl",
    },
    st2: {
      fontFamily: "heading",
      fontWeight: "semibold",
      fontSize: "s",
      lineHeight: "l",
    },
    p1: {
      fontFamily: "body",
      fontWeight: "normal",
      fontSize: "m",
      lineHeight: "xl",
    },
    p2: {
      fontFamily: "body",
      fontWeight: "normal",
      fontSize: "s",
      lineHeight: "l",
    },
    caption: {
      fontFamily: "body",
      fontWeight: "normal",
      fontSize: "xs",
      lineHeight: "m",
    },
    btn1: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "m",
      lineHeight: "xl",
    },
    btn2: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "s",
      lineHeight: "m",
    },
    btn3: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "xs",
      lineHeight: "s",
    },
    btn4: {
      fontFamily: "body",
      fontWeight: "medium",
      fontSize: "2xs",
      lineHeight: "xs",
    },
  },
  defaults: {
    variant: "p1",
  },
};
