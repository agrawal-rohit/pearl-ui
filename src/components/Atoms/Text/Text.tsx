import React from "react";
import { Text as RNText } from "react-native";
import responsiveSize from "../../../utils/responsiveSize";
import { useAtomicComponentConfig } from "../../../hooks/useAtomicComponentConfig";
import {
  backgroundColor,
  BackgroundColorProps,
  color,
  ColorProps,
  layout,
  LayoutProps,
  opacity,
  OpacityProps,
  spacing,
  SpacingProps,
  textShadow,
  TextShadowProps,
  typography,
  TypographyProps,
  visible,
  VisibleProps,
} from "../../../theme/src/styleFunctions";
import {
  ComponentSizes,
  ComponentVariants,
  ResponsiveValue,
  StyleFunctionContainer,
} from "../../../theme/src/types";
import { useTheme } from "../../../hooks/useTheme";

type TextStyleProps = ColorProps &
  BackgroundColorProps &
  OpacityProps &
  VisibleProps &
  LayoutProps &
  TypographyProps &
  SpacingProps &
  TextShadowProps & {
    /** The size of the text */
    size?: ResponsiveValue<ComponentSizes<"Text">>;
    /** The variant of the text */
    variant?: ResponsiveValue<ComponentVariants<"Text">>;
    /** Whether to slightly scale the font size based on the screen dimensions */
    scaleFontSize?: boolean;
  };

export const textStyleFunctions = [
  color,
  backgroundColor,
  opacity,
  visible,
  layout,
  typography,
  spacing,
  textShadow,
] as StyleFunctionContainer[];

type RNTextProps = React.ComponentProps<typeof RNText>;

export type TextProps = TextStyleProps &
  Omit<RNTextProps, keyof TextStyleProps>;

export const buildFontConfig = (
  textStyle: any,
  allowFontScaling: boolean
): object => {
  const fontWeight = textStyle.fontWeight;
  const fontStyle = textStyle.fontStyle || "normal";

  let fontSize = textStyle.fontSize;
  let lineHeight = textStyle.lineHeight;

  if (allowFontScaling) {
    fontSize = responsiveSize(fontSize);
    lineHeight = responsiveSize(lineHeight);
  }

  const initialFontFamily = textStyle.fontFamily;
  const { theme } = useTheme();

  const finalFontFamily =
    theme.fontConfig[initialFontFamily][fontWeight][fontStyle];

  return {
    fontSize,
    fontWeight,
    fontFamily: finalFontFamily,
    fontStyle,
  };
};

/**
 * Text is the component which controls the typography across your app. By default, it renders a <Text /> component
 */
const Text = React.forwardRef((rest: TextProps, ref: any) => {
  const props = useAtomicComponentConfig(
    "Text",
    rest,
    {
      size: rest.size,
      variant: rest.variant,
    },
    "primary",
    textStyleFunctions
  );

  const memoizedBuildFontConfig = React.useCallback(
    () => buildFontConfig(props.style, props.scaleFontSize),
    [props.style, props.scaleFontSize]
  );

  props.style = {
    includeFontPadding: false,
    ...props.style,
    ...memoizedBuildFontConfig(),
  };

  return (
    <RNText
      ref={ref}
      accessible={true}
      accessibilityRole="text"
      allowFontScaling
      {...props}
    >
      {props.children}
    </RNText>
  );
});

export default Text;
