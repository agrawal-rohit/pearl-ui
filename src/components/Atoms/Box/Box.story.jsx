import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "./Box";

storiesOf("Box", module)
  .addDecorator((getStory) => getStory())
  .add("Basic box", () => (
    <Box
      padding="s"
      height={300}
      width={100}
      backgroundColor="primary100"
    ></Box>
  ));
