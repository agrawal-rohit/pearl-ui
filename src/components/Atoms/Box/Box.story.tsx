import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "../Box/Box";

storiesOf("Box", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("Basic box", () => (
    <Box padding="s" width={200} backgroundColor="primary100"></Box>
  ));
