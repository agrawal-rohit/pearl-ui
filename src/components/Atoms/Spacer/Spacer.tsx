import React from "react";
import Box, { BoxProps } from "../box/box";

/** A layout component that creates an adjustable, empty space that can be used to tune the spacing between sibling elements */
const Spacer: React.FC<BoxProps> = ({ children, ...rest }) => {
  return (
    <Box {...rest} flex={1} alignSelf="stretch">
      {children}
    </Box>
  );
};

export default Spacer;
