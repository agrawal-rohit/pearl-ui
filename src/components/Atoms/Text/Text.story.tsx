import React from "react";
import { storiesOf } from "@storybook/react-native";
import Text from "./Text";
import Screen from "../Screen/Screen";
import Box from "../Box/Box";

storiesOf("Typography", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("All", () => (
    <Box flex={1}>
      <Text variant="h1">Heading 1</Text>
      <Text variant="h2">Heading 2</Text>
      <Text variant="h3">Heading 3</Text>
      <Text variant="h4">Heading 4</Text>
      <Text variant="h5">Heading 5</Text>
      <Text variant="h6">Heading 6</Text>
      <Text variant="p1">Paragraph 1</Text>
      <Text variant="p2">Paragraph 2</Text>
      <Text variant="p3">Paragraph 3</Text>
      <Text variant="p4">Paragraph 4</Text>
      <Text variant="label">Label</Text>
      <Text variant="caption">Caption</Text>
      <Box backgroundColor="primary.500">
        <Text variant="btn1">Button 1</Text>
      </Box>
      <Box backgroundColor="primary.500">
        <Text variant="btn2">Button 2</Text>
      </Box>
      <Box backgroundColor="primary.500">
        <Text variant="btn3">Button 3</Text>
      </Box>
      <Box backgroundColor="primary.500">
        <Text variant="btn4">Button 4</Text>
      </Box>
    </Box>
  ));
