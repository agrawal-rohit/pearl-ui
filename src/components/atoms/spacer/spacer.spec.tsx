import React from "react";
import { render } from "@testing-library/react-native";
import { ThemeProvider } from "../../../theme/src/theme-context";
import Spacer from "./spacer";
import Box from "../box/box";
import Text from "../text/text";

jest.useFakeTimers();

describe("Atoms/Spacer", () => {
  it("passes the snapshot test", () => {
    const tree = render(
      <ThemeProvider>
        <Box flexDirection="row">
          <Box p="l" bgColor="danger.400">
            <Text>Box 1</Text>
          </Box>
          <Spacer />
          <Box p="l" bgColor="success.400">
            <Text>Box 2</Text>
          </Box>
        </Box>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
