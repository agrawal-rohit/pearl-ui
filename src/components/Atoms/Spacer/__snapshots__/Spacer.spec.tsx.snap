// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Spacer passes the snapshot test 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "transform": Array [],
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "transform": Array [],
      }
    }
  >
    <View
      animatedStyle={
        Object {
          "value": Object {
            "transform": Array [],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#FF8371",
          "display": "flex",
          "padding": 24,
          "transform": Array [],
        }
      }
    >
      <Text
        accessibilityRole="text"
        accessible={true}
        allowFontScaling={true}
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [],
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "color": "#1A2138",
            "display": "flex",
            "fontFamily": "Poppins-Regular",
            "fontSize": 18.4,
            "fontStyle": "normal",
            "fontWeight": "400",
            "includeFontPadding": false,
            "letterSpacing": 0,
            "lineHeight": 24,
            "transform": Array [],
          }
        }
      >
        Box 1
      </Text>
    </View>
    <View
      animatedStyle={
        Object {
          "value": Object {
            "transform": Array [],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "alignSelf": "stretch",
          "display": "flex",
          "flex": 1,
          "transform": Array [],
        }
      }
    />
    <View
      animatedStyle={
        Object {
          "value": Object {
            "transform": Array [],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#C0E640",
          "display": "flex",
          "padding": 24,
          "transform": Array [],
        }
      }
    >
      <Text
        accessibilityRole="text"
        accessible={true}
        allowFontScaling={true}
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [],
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "color": "#1A2138",
            "display": "flex",
            "fontFamily": "Poppins-Regular",
            "fontSize": 18.4,
            "fontStyle": "normal",
            "fontWeight": "400",
            "includeFontPadding": false,
            "letterSpacing": 0,
            "lineHeight": 24,
            "transform": Array [],
          }
        }
      >
        Box 2
      </Text>
    </View>
  </View>
</View>
`;
